core/src/main/java/hudson/model/labels/LabelExpression.java

```diff
@@ -288,28 +289,35 @@ public static FormValidation validate(@Nullable String expression, @CheckForNull
             for (LabelAtom a : l.listAtoms()) {
                 if (a.isEmpty()) {
                     LabelAtom nearest = LabelAtom.findNearest(a.getName());
-                    return FormValidation.warning(Messages.LabelExpression_NoMatch_DidYouMean(a.getName(),nearest.getDisplayName()));
+                    return FormValidation.warning(Messages.LabelExpression_NoMatch_DidYouMean(a.getName(), nearest.getDisplayName()));
                 }
             }
             return FormValidation.warning(Messages.LabelExpression_NoMatch());
         }
         if (item != null) {
-            // Use the project-oriented validators (including any that might implement checkItem to support non-Project
-            // items too).
-            // FIXME: Perhaps these should aggregate their errors/warnings?
+            final List<FormValidation> problems = new ArrayList<>();
+            // Use the project-oriented validators too, so that validation from older plugins still gets applied.
             for (AbstractProject.LabelValidator v : j.getExtensionList(AbstractProject.LabelValidator.class)) {
                 FormValidation result = v.checkItem(item, l);
-                if (!FormValidation.Kind.OK.equals(result.kind)) {
-                    return result;
+                if (FormValidation.Kind.OK.equals(result.kind)) {
+                    continue;
```

Comment:
BTW I think if https://github.com/jenkinsci/jenkins/blob/339f9b930e13c78c545a2409e016127e36520c87/core/src/main/java/hudson/util/FormValidation.java#L227-L238 were refined slightly, to ignore all occurrences of `FormValidation.OK` in its input, then the API and this impl could be simplified a bit while actually handling `ok(String)` from validators. Not necessary in this PR, just something I noticed while looking at `aggregate`.

Reply:
Yes I looked there too. But I'm not sure it would help to get bulleted list entries for "Label is valid" (assuming a validator might put that in the OK text) among the erroneous ones.

Reply:
Right—`aggregate` would need to be made smarter, perhaps:

* if no arguments, or all `FormValidation.OK`, return `FormValidation.OK`
* if only one non-`FormValidation.OK` argument, return that as is
* else return worst status with bulleted list of messages